receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,539 @@
+type client interface {
+	// GetScalarData retrieves SNMP scalar data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetScalarData(oids []string, processFn processFunc) error
+	// GetIndexedData retrieves SNMP indexed data from a list of passed in OIDS,
+	// then the passed in function is performed on each piece of data
+	GetIndexedData(oids []string, processFn processFunc) error
+	// Connect makes a connection to the SNMP host
+	Connect() error
+	// Close closes a connection to the SNMP host
+	Close() error
+}
+
+// snmpClient implements the client interface and retrieves data through SNMP
+type snmpClient struct {
+	client goSNMPWrapper
+	logger *zap.Logger
+}
+
+// Verify snmpClient implements client interface
+var _ client = (*snmpClient)(nil)
+
+// newClient creates an initialized client
+func newClient(cfg *Config, host component.Host, settings component.TelemetrySettings, logger *zap.Logger) (client, error) {
+	// Create goSNMP client
+	goSNMP := newGoSNMPWrapper()
+	goSNMP.SetTimeout(cfg.CollectionInterval)
+
+	// Set goSNMP version based on config
+	switch cfg.Version {
+	case "v3":
+		goSNMP.SetVersion(gosnmp.Version3)
+	case "v2c":
+		goSNMP.SetVersion(gosnmp.Version2c)
+	case "v1":
+		goSNMP.SetVersion(gosnmp.Version1)
+	default:
+		return nil, errors.New("failed to create goSNMP client: invalid version")
+	}
+
+	// Verify config endpoint is good
+	endpoint := cfg.Endpoint
+	if !strings.Contains(endpoint, "://") {
+		endpoint = "udp://" + endpoint
+	}
+
+	snmpUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goSNMP client: invalid endpoint. %w", err)
+	}
+
+	// Set goSNMP transport based on config
+	switch snmpUrl.Scheme {
+	case "tcp", "tcp4", "tcp6", "udp", "udp4", "udp6":
+		goSNMP.SetTransport(snmpUrl.Scheme)
+	default:
+		return nil, fmt.Errorf("failed to create goSNMP client: unsupported scheme '%s'", snmpUrl.Scheme)
+	}
+
+	// Set goSNMP port based on config
+	portStr := snmpUrl.Port()
+	if portStr == "" {
+		portStr = "161"
+	}
+
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goSNMP client: issue parsing port '%s'. %w", portStr, err)
+	}
+	goSNMP.SetPort(uint16(port))
+
+	// Set goSNMP target based on config
+	goSNMP.SetTarget(snmpUrl.Hostname())
+
+	if goSNMP.GetVersion() == gosnmp.Version3 {
+		// Set goSNMP v3 configs
+		if err := setV3ClientConfigs(&goSNMP, cfg); err != nil {
+			return nil, fmt.Errorf("failed to create goSNMP client: %w", err)
+		}
+	} else {
+		// Set goSNMP community string
+		goSNMP.SetCommunity(cfg.Community)
+	}
+
+	// return client
+	return &snmpClient{
+		client: goSNMP,
+		logger: logger,
+	}, nil
+}
+
+// setV3ClientConfigs sets SNMP v3 related configurations on gosnmp client based on config
+func setV3ClientConfigs(client *goSNMPWrapper, cfg *Config) error {
+	(*client).SetSecurityModel(gosnmp.UserSecurityModel)
```

Comment:
It seems odd to dereference the client pointer here. What's the reasoning behind this?

Reply:
fixed